msaad.go

```diff
@@ -18,97 +18,197 @@ package authaus
-		"officeLocation": null,
-		"preferredLanguage": null,
-		"surname": "Surname",
-		"userPrincipalName": "another@email.address",
-		"id": "5c712197-beef-deef-ffff-f11bb800f365"
-	}
-*/
-type msaadUserJSON struct {
-	DisplayName       string `json:"displayName"`
-	GivenName         string `json:"givenName"`
-	Mail              string `json:"mail"`
-	Surname           string `json:"surname"`
-	MobilePhone       string `json:"mobilePhone"`
-	UserPrincipalName string `json:"userPrincipalName"`
-	ID                string `json:"id"`
-}
-
+	Verbose              bool              // If true, then emit verbose logging
+	DryRun               bool              // If true, don't actually take any action, just log the intended actions
+	TenantID             string            // Your tenant UUID (ie ID of your AAD instance)
+	ClientID             string            // Your client UUID (ie ID of your application)
+	ClientSecret         string            // Secrets used for authenticating Azure AD requests
+	MergeIntervalSeconds int               // If non-zero, then overrides the merge interval
+	DefaultRoles         []string          // Roles that are activated by default if a user has any one of the AAD roles
+	Domain               string            // Domain is embedded in the client configuration as a text field that one can use to scan AD role names and discover permissions. It is only relevant when AutoDiscoverPermissions is set to true
+	RoleToGroup          map[string]string // Map from principleName of AAD role, to Authaus group.
+	AllowArchiveUser     bool              // If true, then archive users who no longer have the relevant roles in the AAD
+	PassthroughClientIDs []string          // Client IDs of trusted IMQS apps utilising app-to-app passthrough auth
+}
+
+// MSAADInterface
+//
+// Interface to abstract the fetching of roles and users, allowing mocking
+// of dependencies (DBs, API functions).
+// Initialize must be called on all implementing structs.
+type MSAADInterface interface {
+	Config() ConfigMSAAD
+	Parent() *Central
+	Provider() MSAADProviderI
+	Initialize(parent *Central, log *log.Logger) error
+	SynchronizeUsers() error
+	IsShuttingDown() bool
+	userBelongsHere(user *msaadUser, matchType MatchType) bool
+	buildCachedRoleGroups() (*cachedRoleGroups, error)
+	syncRoles(roleGroups *cachedRoleGroups, aadUser *msaadUser, internalUserID UserId) error
+	populateAADRoles(users []*msaadUser) error
+	SetConfig(msaad ConfigMSAAD)
+	SetProvider(provider MSAADProviderI)
+}
+
+// MSAADProviderI
+//
+// Interface to abstract the fetching of roles and users, allowing mocking
+// of returns.
+// Initialize must be called on all implementing structs.
+type MSAADProviderI interface {
+	fetchJSON(fetchURL string, jsonRoot interface{}) error
+	doLoggedHTTP(request *http.Request) (*http.Response, []byte, error)
+	doHTTP(request *http.Request) (*http.Response, error)
+	getBearerToken() (token string, expiresAt time.Time, err error)
```

Comment:
The issue is that we need a way to separate e.g. getAADUsers from the calling function (SynchronizeUsers) which is under test...to be able to test SynchronizeUsers, we need a concrete instance of MSAAD, but we don't want to call its own getAADUsers for this, because we want to supply our own...any ideas?